### PR TITLE
issue-997 improve beta_lccdf and beta_lcdf performance

### PR DESCRIPTION
**Pull request by roualdes**
## Summary

Within `stan/math/prim/scal/prob/beta_lccdf.hpp` and `stan/math/prim/scal/prob/beta_lcdf.hpp`, loop around `max_size(alpha, beta)` instead of `N`.

Remove duplicated `using std::exp` in both files above.

## Tests

All tests remain the same and 

`./runTests.py -j3 test/prob/beta` passes.

## Side Effects

None.

## Checklist

- [x] Math issue https://github.com/stan-dev/math/issues/997

- [x] Copyright holder: California State University, Chico

    The copyright holder is typically you or your assignee, such as a university or company. By submitting this pull request, the copyright holder is agreeing to the license the submitted work under the following licenses:
      - Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)
      - Documentation: CC-BY 4.0 (https://creativecommons.org/licenses/by/4.0/)

- [x] the basic tests are passing

    - ~~unit tests pass (to run, use: `./runTests.py test/unit`)~~ `./runTests.py -j3 test/prob/beta` passes
    - header checks pass, (`make test-headers`)
    - docs build, (`make doxygen`)
    - code passes the built in [C++ standards](https://github.com/stan-dev/stan/wiki/Code-Quality) checks (`make cpplint`)

- [x] the code is written in idiomatic C++ and changes are documented in the doxygen

- [x] the new changes are tested
`./runTests.py -j3 test/prob/beta` passes

### CODE REVIEW

File: stan/math/prim/scal/prob/beta_lccdf.hpp [hunk 1]
```diff
@@ -32,14 +32,16 @@ namespace stan {
...
 
 /**
- * Returns the beta log complementary cumulative distribution function for
- * the given probability, success, and failure parameters. Given matching
- * containers returns the log sum of probabilities.
+ * Returns the beta log complementary cumulative distribution function
+ * for the given probability, success, and failure parameters.  Any
+ * arguments other than scalars must be containers of the same size.
+ * With non-scalar arguments, the return is the sum of the log ccdfs
+ * with scalars broadcast as necessary.
```
Comment (bbbales2) [inline stan/math/prim/scal/prob/beta_lccdf.hpp]: I'd prefer "With mixed scalar and container arguments, the scalars are broadcast to containers of length matching the container arguments. With container arguments, this funcion returns the sum of lccdfs."

It makes the broadcast behavior more explicit? As necessary seems vague.

But given this is how stuff is documented everywhere.

I'd be tempted to say just do whatever all the other lccdfs do, but "log sum of probabilities" seems wrong (they're sums of logs of probabilities). Is that why you modified this? Bah, docs are hard.

File: stan/math/prim/scal/prob/beta_lccdf.hpp [hunk 2]
```diff
@@ -32,14 +32,16 @@ namespace stan {
...
  *
  * @tparam T_y type of probability parameter
  * @tparam T_scale_succ type of success parameter
  * @tparam T_scale_fail type of failure parameter
- * @param y probability parameter
+ * @param y type of probability parameter
```
Comment (bbbales2) [inline stan/math/prim/scal/prob/beta_lccdf.hpp]: I think this was a mistake

File: stan/math/prim/scal/prob/beta_lccdf.hpp [hunk 3]
```diff
@@ -96,7 +97,7 @@ typename return_type<T_y, T_scale_succ, T_scale_fail>::type beta_lccdf(
       digamma_sum_vec(max_size(alpha, beta));
 
   if (contains_nonconstant_struct<T_scale_succ, T_scale_fail>::value) {
-    for (size_t i = 0; i < N; i++) {
+    for (size_t i = 0; i < max_size(alpha, beta); i++) {
```
Comment (bbbales2) [inline stan/math/prim/scal/prob/beta_lccdf.hpp]: Good catch!

File: stan/math/prim/scal/prob/beta_lcdf.hpp [hunk 4]
```diff
@@ -33,8 +33,10 @@ namespace math {
 
 /**
  * Returns the beta log cumulative distribution function for the given
- * probability, success, and failure parameters. Given matching containers
- * returns the log sum of probabilities.
+ * probability, success, and failure parameters.  Any arguments other
+ * than scalars must be containers of the same size.  With non-scalar
+ * arguments, the return is the sum of the log ccdfs with scalars
+ * broadcast as necessary.
```
Comment (bbbales2) [inline stan/math/prim/scal/prob/beta_lcdf.hpp]: Same comments as before, but we can just have the convo in one place

**Comment by roualdes** [comment]
Thanks for looking at this.

Most of these doc changes come from a discussion between Bob and I surrounding PR https://github.com/stan-dev/math/pull/975.  The current PR attempts to match what we decided in #975.  I don't have a strong opinion between your suggestion or Bob's previous suggestion.  They both sound good to me.  

To be fair, in that previous PR I should have had `type of probability variate`  not `type of probability parameter` -- though this isn't exactly what your comment is getting at.

Happy to modify as you see fit, as per your comments above or otherwise.  Just let me know.

**Comment by bbbales2** [comment]
Let's roll with what you have for the main doc then. If this has already been covered before, that's good enough for me.

But y is not a type, and it's not necessarily a probability. It's a number between zero and one, sure.

How about:

```
 * @tparam T_y type of y
 * @tparam T_scale_succ type of success parameter
 * @tparam T_scale_fail type of failure parameter
 * @param y (Sequence of) scalar(s) between zero and one
 * @param alpha (Sequence of) success parameter(s)
 * @param beta (Sequence of) failure parameter(s)
```

for the arguments. Like https://github.com/stan-dev/math/blob/7ea12d4947ec28638d4f4957d92cb92720a3e909/stan/math/prim/scal/prob/std_normal_lpdf.hpp#L27 ?

And that kindof vectorizes the doc as well. If that seems good update the other one too. Hoooray for seemingly innocent doc changes hahaha.

**Comment by bbbales2** [comment]
Also props for being pull req. #1000

**Comment by bob-carpenter** [comment]
Ben's suggestion on the vectorized grammar is great, and I'm super picky on that front.

Statistically, alpha is the prior success count plus one and beta the prior failure count plus one, but they are parameters and they're continuous, so what Ben wrote is OK.

**Comment by bob-carpenter** [comment]
Indeed!  That's a scary number of PRs given that I think I've at least read most of them.

We have to wait for 5K on Stan itself---everything started there before it was moved, so it's artificially ahead.

**Pull request merged**